### ISSUE TITLE
Add the current version to the AST (Root)

### ISSUE THREAD

**Issue by michael-ciniawsky** [postcss/postcss#1359]
Would it be possible to add the current version of `postcss` or ideally just a `{Number}` to e.g `ast.root`? So third party tools passing the AST can check for compatibility between versions of `postcss` (AST's).

```js
import postcss from 'postcss'

const ast = postcss.root()
```

```diff
Root {
  type: 'root',
  raws: {},
  nodes: [],
+ version: 7  // postcss.pkg.version
}
```

What is the current policy for changes to the AST, assuming it is a always BREAKING CHANGE seems the most logical I think?

https://github.com/postcss/postcss-loader/pull/425 for context on why this maybe needed

**Comment by ai** [comment]
@michael-ciniawsky you can have access to the PostCSS version in `result.processor.version`.

What is a use case to have it in AST?

**Comment by michael-ciniawsky** [comment]
Ahh I knew version was already avaiable somewhere but I forgot whwre ^=^ I will check it out just a few moments this maybe enough. The issue popped up in `postcss-loader` since loaders before and after `postcss-loader` can theoretically reuse the AST instead of reparsing the source string, but we need to be sure that there are no incompatibilities between `postcss` versions used by different loaders

**Comment by michael-ciniawsky** [comment]
Is there a way to get the current version before processing the CSS/AST in some way? What we try to accomplish is the following (simplified).

```js
import postcss from 'postcss'

function loader (src, map, meta) {
  const ast = null

  if (meta.ast.version === postcss.version) {
     ast = meta.ast.root
  }

  postcss(plugins).process(ast || src, options)...

  ...
}
```

This currently requires to `require(postcss/package)` to get the version of `postcss` used and either a utility or the full `semver` package to compare the AST with the `postcss` version used, which I believe should be unnecessary if there is a way to explicitly compare an input AST with the currently supported AST version of the given `postcss` version somehow. Otherwise each third-party package needs to add it's own way of handling this case so my thinking was maybe there might be an easier way to do this. It's also not always the case that `ast.version` => `postcss.version` since not every postcss `major` may include breaking chance to the PostCSS AST which adds additional complexity. E.g something in this direction should be sufficient here

```js
import postcss from 'postcss'

function loader (src, map, meta) {
  const ast = null
   
  console.log(meta.ast.version) // => 7
  console.log(postcss.ast.version) // => 7

  if (meta.ast.version === postcss.ast.version) {
     ast = meta.ast.root
  }

  postcss(plugins).process(ast || src, options)...

  ...
}
```

**Comment by michael-ciniawsky** [comment]
It's not an pressing issue, but before `postcss-loader` handles it in it's own way I wanted to check if there is general interest in solving this (maybe it's relevant to other tool chains aswell) so feel free to close

**Comment by ai** [comment]
Adding `Root#version` is not hard. But there is one problem :).

AST can contain parts of different AST versions because plugins use own `postcss` version to create AST nodes.

We can add `Node#version` or `Node#[postcss.version]` (`postcss.version` will be a Symbol).

**Comment by michael-ciniawsky** [comment]
`Root#version`  and `postcss.version` would be awesome in my opinion as this avoids loading the `package.json` multiple times.

Ideally the `Root#version` is indenpendent of `postcss.version` and a simple `{Number}` to also avoid parsing the semver string here. E.g

```js
// Need to parse the semver string e.g with `semver`
if (semver.satisfies(prevAST.root.version, postcss.root.version || postcss.version) {

}

// {Number} === {Number}
if (prevAST.root.version === postcss.root.version) {

}
```

On the other hand `postcss.version`  should contain the semver string to e.g check for features in minor release (for plugins) etc, maybe also a good idea to provided it already parsed (alongside the string) e.g

```js
import postcss from 'postcss'

const NAME = 'my-plugin'
// All of this is a bit ugly there might be cleaner ways to to this checks
// but shoud be fine for explanation purposes 
postcss.plugin(NAME, (options) => {
   return (root, result) => {
         const processor = result.processor

         if (postcss.version.major < processor.version.major) {
             result.warn(`
               ${NAME} needs a minimum postcss version of ${postcss.version.toString()}, 
               but the current version is ${processor.version.toString()}`. 
               Therefore the plugin was skipped and didn't run
             `)

             return
         }

         if (postcss.version.minor >= processor.version.minor) {
             root.useNewFeature()
             node.useNewFeature()
         }

         // etc...
   }
}
```

So in summary the following might be a good direction to solve versioning problems in the future

```js
console.log(postcss.version) // => e.g '7.0.30'
console.log(ast.root.version) // => e.g 7
console.log(node.version) // => e.g 7 (=== ast.root.version)
```

My only concern with adding `Node#version` is increased memory usage especially for big CSS Files as this adds up quickly when working with AST's. But maybe acceptable depending on how common it is for plugins and how big of an concern this is, but can't really comment here...

**Comment by ai** [comment]
We can’t use just a number, because in the minor version we can add a method to AST

**Comment by ai** [comment]
> My only concern with adding Node#version is increased memory usage especially for big CSS Files as this adds up quickly when working with AST's. But maybe acceptable depending on how common it is for plugins and how big of an concern this is, but can't really comment here...

It bothers me too. Do we really need this version?

In that “Gatsby” PR we do not need `root.version`.

**Comment by michael-ciniawsky** [comment]
The issue in `postcss-loader` simply needs a way to compare version of an input AST to the version of `postcss` used by the loader, before calling `.process(ast, options)` to avoid incompatible AST's between versions of  `postcss`.

**Comment by ai** [comment]
> The issue in postcss-loader simply needs a way to compare version of an input AST to the version of postcss used by the loader, before calling .process(ast, options) to avoid incompatible AST's between versions of postcss.

PostCSS has own warning about incompatibility in the case of error.

Do we need an additional check?

**Comment by michael-ciniawsky** [comment]
Just `Root#version`  and `postcss.version` would solve the issue. E.g 

```js
// Where root is a previous.ast or result.ast
root.version => '7.0.30'
postcss.version => '7.0.30'

// Should be fine to compare the strings
if (root.version === postcss.version) {
  postcss(plugins).process(root || css, options)...
}
```

**Comment by ai** [comment]
I am closing this issue for now, because I do not have time (too many open-source projects).

I can look at this idea again if somebody give me good review and popular use cases.
